influence_simulator/influence_simulator.py

```diff
@@ -0,0 +1,139 @@
+        self.random_state = random_state
+
+        self.result: list[SimulationResult] = []
+
+    def init_simulation(self, node: int) -> IndependentCascadeModel:
+        model = IndependentCascadeModel(
+            graph=self.graph,
+            initially_infected=0,
+            infection_probability=self.infection_probability,
+            random_state=self.random_state,
+        )
+
+        model.state[node] = "infected"
+        return model
+
+    def simulate_node(self, node: int) -> SimulationResult:
+        model = self.init_simulation(node)
+
+        peak = Peak(0, 0)
+        while not model.terminated:
+            model.step()
+
+            if (infected := model.state_summary["infected"]) > peak.infected:
+                peak = Peak(infected, model.iteration)
+
+        final_state = model.state_summary
+        return SimulationResult(
+            node=node,
+            simulation_length=model.iteration,
+            exposed=final_state["recovered"],
+            not_exposed=final_state["susceptible"],
+            peak_infected=peak.infected,
+            peak_iteration=peak.iteration,
+        )
+
+    def simulate(
+        self,
+        *,
+        n_jobs: int = 1,
+        chunksize: int = 1,
+        verbose: bool = True,
+    ) -> None:
+
+        if n_jobs == 1:
+            if verbose:
+                nodes = tqdm(self.graph.nodes)
+            else:
+                nodes = self.graph.nodes
+
+            self.result = [self.simulate_node(node) for node in nodes]
+        else:
+            if n_jobs <= 0:
+                n_jobs = None
+
+            self.result = process_map(
+                self.simulate_node,
+                self.graph.nodes,
+                chunksize=chunksize,
+                max_workers=n_jobs,
+                disable=not verbose,
+            )
+
+    @staticmethod
+    def get_loading_function(path: str) -> Callable:
+        extension = path.split(".")[-1]
+        try:
+            func = getattr(nx, f"read_{extension}")
+        except AttributeError as e:
+            raise ValueError(
+                f"Graph format '{extension}' not supported."
+            ) from e
+
+        return func
+
+    @staticmethod
+    def relabel_nodes(graph: nx.Graph) -> nx.Graph:
+        """
+        Changes type of node labels to ints.
+        '0' -> 0
+        """
+        mapping = {}
+        for i in range(len(graph)):
+            mapping[str(i)] = i
+
+        return nx.relabel_nodes(graph, mapping)
+
+    @staticmethod
+    def load_graph(path: str) -> nx.Graph:
+        loading_function = InfluenceSimulator.get_loading_function(path)
+        graph = loading_function(path)
+
+        if isinstance(graph, nx.MultiGraph):
+            graph = nx.Graph(graph)
+
+        graph = InfluenceSimulator.relabel_nodes(graph)
+        return graph
+
+    def save_result(self, path: str) -> None:
+        with open(path, "wt", encoding="utf-8") as handle:
```

Comment:
May be some check whether Is there any result to save?

Reply:
Done